### PR TITLE
Add support for `afterTransform` hook

### LINKED ISSUE #14: Add visitor function
### Initial checklist

- [X] I read the [support docs](https://github.com/syntax-tree/.github/blob/main/support.md)
- [X] I read the [contributing guide](https://github.com/syntax-tree/.github/blob/main/contributing.md)
- [X] I agree to follow the [code of conduct](https://github.com/syntax-tree/.github/blob/main/code-of-conduct.md)
- [X] I searched [issues](https://github.com/search?q=user%3Asyntax-tree&type=Issues) and couldn’t find anything (or linked relevant results below)

### Problem

Really like the *unified* ecosystem! I'm currently working on a [MutationObserver](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver)-based project where those DOM mutations need to be serialized and deserialized. To reconstruct a DOM mutation (e.g., changing the value of a text node), I need to keep track of all DOM nodes that have been deserialized by `hast-util-to-dom` and connect those to the appropriate hast node and vice-versa. Overview for context:

Record (not relevant for this feature request)

1. Take a snapshot of the whole DOM and serialize it to a virtual DOM (hast)
2. MutationObserver detects a text change: [listener](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver/observe#parameters) contains the target DOM node, and with the help of a custom node store, everything gets tracked and serialized
 
Replay

3. Deserialize the entire DOM snapshot with `hast-util-to-dom`
4. Identify the target DOM node and apply the text change: Unfortunately, this step is not possible with the current `hast-util-to-dom` functionality because there's no reliable way to map a generated DOM node to a hast node.

### Solution

By adding a visitor function, it's possible to inject custom logic to the tree traversal to accomplish the DOM<>VDOM mapping needed in the above example.

```
options.onNode: (hastNode: HastNode, domNode: Node) => void
```

Usage example:

```js
function deserialize(
  // A custom store to track all DOM<>VDOM associations
  store: NodeStore,
  // The virtual DOM to deserialize
  vdom: HastNode,
): Node {
  return toDom(vdom, {
    // The new visitor function that receives both nodes as arguments
    onNode: (hastNode: HastNode, domNode: Node) => {
      // Custom code here...
      store.addNode(domNode, hastNode);
    },
  });
}
```

Although not required for my use case, such a visitor function could also be used in the future similarly to [unist-util-visit-parents#visitor](https://github.com/syntax-tree/unist-util-visit-parents#next--visitornode-ancestors) to transform nodes while the tree is being traversed instead of looping multiple times through the entire tree.

### Alternatives

There's currently no viable alternative. What I've considered for my specific use case:
* Add a `data-node-id` attribute to the DOM node, which gets included in HastElement.properties. However, this is only possible for (Hast)Element nodes. Text and many other DOM nodes don't have attributes.
* A selector-based approach to identify the nodes (e.g., with XPath) is not an option because DOM nodes could have been added or removed, breaking the selectors.

Happy to provide a PR with the addition!

### PR DESCRIPTION
**Pull request by danistefanovic**
### Initial checklist

*   [x] I read the support docs 
*   [x] I read the contributing guide 
*   [x] I agree to follow the code of conduct 
*   [x] I searched issues and couldn’t find anything (or linked relevant results below) 
*   [x] If applicable, I’ve added docs and tests

### Description of changes

Closes #14

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -88,6 +88,10 @@ Document interface to use (default: `globalThis.document`).
 
 `namespace` to use to create [*elements*][element].
 
+###### `options.afterTransform(node, transformed)`
+
+Callback invoked after each node transformation (`Function`).
```
Comment (wooorm) [inline readme.md:93]:
```suggestion
###### `options.afterTransform`

Function called when a hast node is transformed into a DOM node (`Function?`).
Given the hast node that was handled as the first parameter and the
corresponding DOM node as the second parameter.
```

How about this?

File: lib/index.js [hunk 2]
```diff
@@ -12,12 +12,14 @@
  * @property {boolean} [fragment=false] Whether a DOM fragment should be returned
```
Comment (wooorm) [inline lib/index.js:12]: Good to expose a type of the function, something like this:

```
* @typedef AfterTransform
*   Function called when a hast node is transformed into a DOM node.
* @param {HastNode} hastNode
*   The hast node that was handled
* @param {DomNode} domNode
*   The corresponding DOM node
* @returns {void}
```

File: lib/index.js [hunk 3]
```diff
@@ -12,12 +12,14 @@
...
  * @property {Document} [document] Document interface to use (default: `globalThis.document`)
  * @property {string} [namespace] `namespace` to use to create elements
+ * @property {(node: HastNode, transformed: Node) => void} [afterTransform] Callback invoked after each node transformation
```
Comment (wooorm) [inline lib/index.js:15]:
```suggestion
 * @property {AfterTransform} [afterTransform] Function called when a hast node
 *   is transformed into a DOM node
```

File: lib/index.js [hunk 4]
```diff
@@ -12,12 +12,14 @@
...
  *
  * @typedef Context
  * @property {Document} doc
  * @property {boolean} [fragment=false]
  * @property {string} [namespace]
  * @property {string} [impliedNamespace]
+ * @property {Options['afterTransform']} [afterTransform]
```
Comment (wooorm) [inline lib/index.js:22]:
```suggestion
 * @property {AfterTransform} [afterTransform]
```

File: lib/index.js [hunk 5]
```diff
@@ -30,19 +32,28 @@ import {find, html, svg} from 'property-information'
...
+        return text(node, ctx)
+      case 'element':
+        return element(node, ctx)
+      case 'doctype':
+        return doctype(node, ctx)
+      case 'comment':
+        return comment(node, ctx)
+      default:
+        return element(node, ctx)
+    }
```
Comment (wooorm) [inline lib/index.js:56]: No need for the function in a function?
Either call the original `transform` something like `one` or `handle` and use it in the new transform, or change the switch to instead place the correct root, text, element, ... function in a `handle` variable and call it?

**Review by wooorm — COMMENTED** [review]
Thanks for the PR. The result is good! I think the code, types, and docs can improve a bit:

File: lib/index.js (hunk 5 above)
Comment (danistefanovic) [reply lib/index.js]: Agreed 👍 I was trying to mimic the code style from other syntax-tree repositories.

**Comment by wooorm** [comment]
Thanks Daniel, released!

**Comment by danistefanovic** [comment]
Awesome. Thanks!

**Pull request merged**